### PR TITLE
"... has no deployed releases" error when release history contains only failed releases and history limit reached

### LINKED ISSUE #10121: "... has no deployed releases" error when release history contains only failed releases and history limit reached
An error will arise when there are no successfully deployed release yet, but releases history limit has been reached. In such situation helm will refuse to upgrade release anymore with "... has no deployed releases" error.

Output of `helm version`:

```
version.BuildInfo{Version:"v3.6.3", GitCommit:"d506314abfb5d21419df8c7e7e68012379db2354", GitTreeState:"clean", GoVersion:"go1.16.5"}
```

### PR DESCRIPTION
**Pull request by distorhead**
Fixed old-releases rotation procedure to not require a deployed release to exists.

An error will arise when there are no successfully deployed release yet, but releases history limit has been reached. In such situation helm will refuse to upgrade release anymore with "... has no deployed releases" error.

Furthermore, release rotation procedure already expecting lastDeployedRelease to be either nil, or not nil. So it is assumed that deployed release may exist or may not and these both outcomes were already expected as a valid situation rather than a failure.

### CODE REVIEW

**Comment by distorhead** [comment]
Not sure what to do with this test case: https://github.com/helm/helm/commit/8c28da65676a190623ac1d10711780e58e574a04#diff-82890bcea98fd02fb22d71e2add4a8536950a0198fa2e604b73c5882e3ef2d94R300

I see initial motivation to add this test case was to "check that removeLeastRecent error will fail release.Create procedure": https://github.com/helm/helm/commit/00cf10d360de3fbe440789ee51662c2894e041ce#diff-8b98b331c5d4acbeb7274c68973d20900daaed47c8d8f3e62ba39284379166bbL64

So removeLeastRecent procedure could raise any error just for error-handling check. I have removed ErrNoDeployedReleases, which has been used as an example error, which removeLeastRecent function could produce. Not sure yet what "example-error" we could use for this test case as a replacement for ErrNoDeployedReleases.

**Comment by yxxhero** [comment]
Can you add some unittest for this change? @distorhead  Thanks very much.

**Comment by distorhead** [comment]
> Can you add some unittest for this change? @distorhead Thanks very much.

@yxxhero Fixed test case.

**Comment by zonggen** [comment]
It would be great if you create an issue associated to this PR and gets triaged as `bug`([link to contribution procedure](https://github.com/helm/helm/blob/main/CONTRIBUTING.md#how-to-contribute-a-patch)).

File: pkg/storage/storage_test.go
Comment (zonggen) [inline pkg/storage/storage_test.go:281]: This variable seems like having more visibility than it needs.
```suggestion
var errMaxHistoryMockDriverSomethingHappened = errors.New("something happened")
```

Comment (distorhead) [reply pkg/storage/storage_test.go:281]: Thanks!

**Comment by distorhead** [comment]
Added issue: https://github.com/helm/helm/issues/10121
Added new test case for issue.

**Review by bacongobbler — APPROVED** [review]
LGTM. Thanks. Will need another maintainer's approval before merging.

**Review by hickeyma — COMMENTED** [review]
@distorhead Thanks for this. Looks good when tested manually. Just have some questions on UT.

File: pkg/storage/storage_test.go
Comment (distorhead) [reply pkg/storage/storage_test.go]: Hi! One more release will be created here: https://github.com/helm/helm/pull/10085/files#diff-82890bcea98fd02fb22d71e2add4a8536950a0198fa2e604b73c5882e3ef2d94R515

Changed status of 5th release also to StatusFailed. It was just my initial example case to reproduce an error. It does not matter much, but may confuse.

**Comment by distorhead** [comment]
@hickeyma Hello, is there any problem with this PR that should be fixed?

**Comment by hickeyma** [comment]
@distorhead Seems to be an issue with circle build. do you mins pouching again to kick off a build?

File: pkg/storage/storage_test.go
Comment (hickeyma) [reply pkg/storage/storage_test.go:493]: ok

**Review by hickeyma — APPROVED** [review]

**Comment by hickeyma** [comment]
Merging this without CI as it seemed to be locked and unable to start it.

**Pull request merged**
